beacon-chain/blockchain/van_minimal_consensus_info.go

```diff
@@ -0,0 +1,197 @@
+package blockchain
+
+import (
+	"encoding/hex"
+	"fmt"
+	types2 "github.com/gogo/protobuf/types"
+	types "github.com/prysmaticlabs/eth2-types"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// MinimalConsensusInfoFetcher retrieves the minimal consensus info for provided epoch from blockchain service
+type MinimalConsensusInfoFetcher interface {
+	MinimalConsensusInfo(types.Epoch) (*ethpb.MinimalConsensusInfo, error)
+	MinimalConsensusInfoRange(types.Epoch) ([]*ethpb.MinimalConsensusInfo, error)
+}
+
+//MinimalConsensusInfo provides minimalConsensusInfo for one epoch provided as an argument
+func (s *Service) MinimalConsensusInfo(epoch types.Epoch) (minConsensusInfo *ethpb.MinimalConsensusInfo, err error) {
+	log.WithField("prefix", "GetPastMinimalConsensusInfo").WithField("epoch", uint64(epoch))
+
+	assignments, err := s.getPastProposerListForEpoch(epoch)
+	if nil != err {
+		log.Errorf("[VAN_SUB] getProposerListForEpoch err = %s", err.Error())
+		return nil, err
+	}
+
+	assignmentsSlice := make([]string, 0)
+
+	// Slot 0 was never signed by anybody
+	if 0 == epoch {
+		publicKeyBytes := make([]byte, params.BeaconConfig().BLSPubkeyLength)
+		currentString := fmt.Sprintf("0x%s", hex.EncodeToString(publicKeyBytes))
+		assignmentsSlice = append(assignmentsSlice, currentString)
+	}
+
+	for _, assigment := range assignments.Assignments {
+		currentString := fmt.Sprintf("0x%s", hex.EncodeToString(assigment.PublicKey))
+		assignmentsSlice = append(assignmentsSlice, currentString)
+	}
+
+	expectedValidators := int(params.BeaconConfig().SlotsPerEpoch)
+
+	if len(assignmentsSlice) != expectedValidators {
+		err := fmt.Errorf(
+			"not enough assignments, expected: %d, got: %d",
+			expectedValidators,
+			len(assignmentsSlice),
+		)
+		log.Errorf("[VAN_SUB] Assignments err = %s", err.Error())
+
+		return nil, err
+	}
+
+	genesisTime := s.genesisTime
+	startSlot, err := helpers.StartSlot(epoch)
+	if nil != err {
+		log.Errorf("[VAN_SUB] StartSlot err = %s", err.Error())
+		return nil, err
+	}
+	epochStartTime, err := helpers.SlotToTime(uint64(genesisTime.Unix()), startSlot)
+	if nil != err {
+		log.Errorf("[VAN_SUB] SlotToTime err = %s", err.Error())
+		return nil, err
+	}
+
+	minConsensusInfo = &ethpb.MinimalConsensusInfo{
+		Epoch:            epoch,
+		ValidatorList:    assignmentsSlice,
+		EpochTimeStart:   uint64(epochStartTime.Unix()),
+		SlotTimeDuration: &types2.Duration{Seconds: int64(params.BeaconConfig().SecondsPerSlot)},
+	}
+
+	log.Infof("[VAN_SUB] currEpoch = %#v", uint64(epoch))
+
+	return minConsensusInfo, nil
+}
+
+//MinimalConsensusInfoRange provides minimalConsensusInfo for all epochs from epoch provided as an argument
+func (s *Service) MinimalConsensusInfoRange(
+	fromEpoch types.Epoch,
```

Comment:
If from epoch is far future from current epoch then what will happen? I could not see this check? Should be a checking condition in here:
```go
        currentEpoch := helpers.SlotToEpoch(bs.GenesisTimeFetcher.CurrentSlot())
	if from > currentEpoch {
		return nil, status.Errorf(
			codes.InvalidArgument,
			errEpoch,
			currentEpoch,
			fromEpoch,
		)
	}
```

Reply:
You're right. I've prepared this check as you has suggested, as it is in `ListValidatorAssignments` too. I've managed to reimplement this fragment too then:
```
minimalConsensusInfoRange, err := bs.MinimalConsensusInfoRange(bs.Ctx, requestedFromEpoch)
	if err != nil {
		return status.Errorf(codes.Unavailable, "Could not get MinimalConsensusInfoRange for a stream: %v", err)
	}
```

Reply:
I've decide to move it to the upper level:
```
currentEpoch := helpers.SlotToEpoch(bs.GenesisTimeFetcher.CurrentSlot())
	if requestedFromEpoch > currentEpoch {
		return status.Errorf(
			codes.InvalidArgument,
			errEpoch,
			currentEpoch,
			requestedFromEpoch,
		)
	}

	minimalConsensusInfoRange, err := bs.MinimalConsensusInfoRange(bs.Ctx, requestedFromEpoch)
	if err != nil {
		return status.Errorf(codes.Unavailable, "Could not get MinimalConsensusInfoRange for a stream: %v", err)
	}
```